File: mobile/services/booking.ts
import api from './api';

/**
 * Booking API calls
 */

export interface Location {
  address: string;
  latitude: number;
  longitude: number;
}

export interface CreateBookingRequest {
  pickup_location: Location;
  destination: Location;
  distance: number;
  estimated_time: number;
  base_fare: number;
  gratuity: number;
  total_fare: number;
  notes?: string;
}

export interface Booking {
  id: string;
  user_id: string;
  driver_id?: string;
  pickup_location: Location;
  destination: Location;
  distance: number;
  estimated_time: number;
  base_fare: number;
  gratuity: number;
  total_fare: number;
  notes: string;
  status: 'pending' | 'accepted' | 'in_progress' | 'completed' | 'cancelled';
  created_at: string;
  updated_at: string;
}

export interface BookingListResponse {
  bookings: Booking[];
  total: number;
}

/**
 * Create a new ride booking
 */
export async function createBooking(data: CreateBookingRequest): Promise<Booking> {
  const response = await api.post<Booking>('/bookings/create', data);
  return response.data;
}

/**
 * Get current user's booking history
 */
export async function getMyBookings(limit: number = 20, skip: number = 0): Promise<BookingListResponse> {
  const response = await api.get<BookingListResponse>('/bookings/my-bookings', {
    params: { limit, skip },
  });
  return response.data;
}

/**
 * Get specific booking details
 */
export async function getBooking(bookingId: string): Promise<Booking> {
  const response = await api.get<Booking>(`/bookings/${bookingId}`);
  return response.data;
}

/**
 * Cancel a booking
 */
export async function cancelBooking(bookingId: string): Promise<{ message: string }> {
  const response = await api.delete<{ message: string }>(`/bookings/${bookingId}`);
  return response.data;
}

File: backend/app/schemas/booking.py
"""
Booking request/response schemas
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class LocationSchema(BaseModel):
    """Location coordinates and address"""
    address: str
    latitude: float
    longitude: float


class CreateBookingRequest(BaseModel):
    """Create new booking request"""
    pickup_location: LocationSchema
    destination: LocationSchema
    distance: float = Field(..., description="Distance in kilometers")
    estimated_time: int = Field(..., description="Estimated time in minutes")
    base_fare: float = Field(..., description="Base fare amount")
    gratuity: float = Field(default=0, description="Gratuity/tip amount")
    total_fare: float = Field(..., description="Total fare (base + gratuity)")
    notes: Optional[str] = Field(default="", description="Notes for driver")


class BookingResponse(BaseModel):
    """Booking response"""
    id: str
    user_id: str
    driver_id: Optional[str] = None
    pickup_location: LocationSchema
    destination: LocationSchema
    distance: float
    estimated_time: int
    base_fare: float
    gratuity: float
    total_fare: float
    notes: str
    status: str  # pending, accepted, in_progress, completed, cancelled
    created_at: datetime
    updated_at: datetime


class BookingListResponse(BaseModel):
    """List of bookings for user"""
    bookings: list[BookingResponse]
    total: int


class UpdateBookingStatusRequest(BaseModel):
    """Update booking status"""
    status: str = Field(..., description="New status: accepted, in_progress, completed, cancelled")
    driver_id: Optional[str] = Field(default=None, description="Driver ID if accepting")
